harvard-lts/spotlight-oaipmh-resources
# Spotlight 3.3.0 changes

# Summary

Updates this gem to be compatible with [Spotlight 3.3.0](https://github.com/projectblacklight/spotlight/tree/v3.3.0) (from 2.7.0). 

There are several tickets relevant to this PR: 
- https://github.com/harvard-lts/CURIOSity/issues/13 
- https://github.com/harvard-lts/CURIOSity/issues/17 
- https://github.com/harvard-lts/CURIOSity/issues/21

## Harvesting 

_https://github.com/harvard-lts/CURIOSity/issues/13_ 

Correctly imports resources and relies on sidecar auto creation to properly generate the sidecars.

Notable changes: 

- Convert `Spotlight::Resources::OaipmhHarvester` to `Spotlight::OaipmhHarvester` 
  - Spotlight has overhauled how they index resources, and the harvester being a resource was causing issues 
- Broke up harvesting logic into sub-methods 
- Move OAI scraping logic from `PerformHarvestsJob` to harvester model
- Move all OAI parsing logic to `OaipmhModsParser` (was `OaipmhModsItem`) 
- Create new `OaipmhUpload` resource to be an ActiveRecord mirror of the data in Solr 
- Extract all used logic from `OaipmhBuilder` and delete class 
  - Spotlight uses an ETL pipeline now instead of builders 

## JobTracking

_https://github.com/harvard-lts/CURIOSity/issues/17_ 

Notable changes: 

- Add Spotlight's built-in `JobTracking` logic to `PerformHarvestsJob` 

### Screenshots 

<details><summary>Harvesting shows up in Dashboard > Indexing activity</summary>

![Curation - Dashboard dev - Harvard Curiosity 2022-07-21 at 10 50 40 AM](https://user-images.githubusercontent.com/32469930/180281348-5294b43e-638f-457d-bee5-2bf4a34b0a36.jpg)

</details>

<details><summary>Harvesting progress tracking</summary>

![Curation - Dashboard dev - Harvard Curiosity 2022-07-21 at 10 51 26 AM](https://user-images.githubusercontent.com/32469930/180281382-c4f504b4-1d09-439f-b430-d31537efc4b3.jpg)

</details>

## Permalinks / URNs

_https://github.com/harvard-lts/CURIOSity/issues/21_

- Replace the exhibit id & generic spotlight ID in the slug  for harvested items with the `external_id` (a column that includes both the exhibit id and the URN)
- Make the external ID uppercase so that it matches with the standard way of writing URNs (they are all uppercase now)
- When exhibits are harvested, items should all come in with an uppercase URN, regardless of if their IDs were saved as lowercased or not. 

## Miscellaneous

- Add configuration options 
  - `download_full_image` 
  - `standard_spotlight_fields` 
  - `use_iiif_images` 
- Changes to engine to allow gem migrations to work correctly in applications 
- Update specs 

# Testing Instructions 

1. In CURIOSity, `git checkout harvesting-take-3` 
2. Login as an admin 
3. Navigate to `/spotlight/new` and create a new Exhibit 
4. Click "Exhibit Dashboard" from the dropdown at the top right of the navbar 
5. Click "Items" in the left-hand sidebar 
6. Click "Add Items" 
7. Fill out the form with the following credentials 
  a. Base URL - https://api.lib.harvard.edu/oai/
  b. Set spec - demo
  c. Mapping File - upload [this file](https://github.com/harvard-lts/spotlight-oaipmh-resources/blob/job_entry/harvard_yaml_mapping_files/mods/curiosity.yml) 
8. Click "Add MODS Items" 
9. Click "Dashboard" from the left-hand sidebar 
10. Track the progress of the harvest under the "Recent item indexing activity" section 
11. Once the harvest finishes, verify that items show up in the exhibit (e.g. via the search results) 

### Interested parties: @dl-maura @phil-plencner-hl
